Add local caching for remote actor avatars

pfefferle opened the pull request:
fixes #926

## Proposed changes:

* Cache remote actor avatars locally to reduce external dependencies and improve load times
* Download and optimize avatars when actors are created or updated
* Reuse the image optimization from #2609 (resize to 1200px, convert to WebP)
* Track remote avatar URL separately to detect changes and avoid unnecessary re-downloads
* Clean up cached avatar files when actor is deleted
* Gracefully fall back to remote URL if local caching fails

### Other information:

- [x] Have you written new tests for your changes, if applicable?

Uses existing test coverage - the avatar caching gracefully handles test environments where HTTP downloads aren't mocked.

**Note:** This PR depends on #2609 (image optimization) - set as base branch.

## Testing instructions:

* Set up a local WordPress environment with wp-env
* Follow a remote ActivityPub actor (e.g., from Mastodon)
* Check `wp-content/uploads/activitypub/actors/{actor_id}/` for the cached avatar
* The avatar should be resized and converted to WebP format
* Verify the avatar displays correctly in followers/following lists
* Update the remote actor's avatar on their server
* Wait for the scheduled update (or manually trigger `activitypub_update_remote_actors`)
* Verify the new avatar is downloaded and cached

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [x] Added - for new features
-   [ ] Changed - for changes in existing functionality
-   [ ] Deprecated - for soon-to-be removed features
-   [ ] Removed - for now removed features
-   [ ] Fixed - for any bug fixes
-   [ ] Security - in case of vulnerabilities

#### Message

Add local caching for remote actor avatars.

</details>

pfefferle committed:
Add image optimization for imported attachments

Optimize imported ActivityPub images by resizing large images and
converting to WebP format for reduced file sizes:

- Add MAX_IMAGE_DIMENSION constant (1200px)
- Add optimize_image() method using WordPress image editor
- Convert images to WebP when supported, fallback to JPEG
- Skip GIFs (may be animated) and SVGs (vector graphics)
- Update save_file() and save_attachment() to optimize images
- Update tests to handle WebP conversion

matticbot committed:
Add changelog

pfefferle committed:
Fix optimize_image logic for edge cases

Address Copilot feedback:
- Early return when no resize needed for PNG/WebP (avoids returning true)
- Simplify result handling - editor->save() always returns array
- Set changed=true when any save operation occurs

pfefferle committed:
Add local caching for remote actor avatars

Cache avatars from remote actors locally to reduce external dependencies
and improve load times:

- Add actors directory constant and cache_actor_avatar method
- Download and optimize avatars on actor create/update
- Resize avatars to 100px max (MAX_AVATAR_DIMENSION constant)
- Store local URL in _activitypub_avatar_url meta
- Clean up avatar files when actor is deleted
- Fall back to remote URL if local caching fails

pfefferle committed:
Add max_dimension parameter to optimize_image method

Allow callers to specify the maximum dimension for image resizing.

pfefferle committed:
Refactor avatar caching to align with Attachments class patterns

- Use get_storage_paths() for actor avatar storage paths.
- Rename cache_actor_avatar to save_actor_avatar.
- Rename delete_actor_avatar_files to delete_actors_directory.
- Remove separate on_delete hook wrapper function.
- Add 'actor' type support to get_storage_paths().

pfefferle committed:
Add max_dimension parameter to save_file method

Allow callers to specify custom image dimensions when saving files,
enabling reuse for different contexts like avatars.

pfefferle committed:
Add tests for save_actor_avatar and delete_actors_directory

- Test successful avatar save with valid URL.
- Test empty and invalid URL handling.
- Test download error handling.
- Test avatar replacement.
- Test directory deletion.
- Test non-actor post type protection.

pfefferle committed:
Add tests for save_actor_avatar and delete_actors_directory

- Test successful avatar save with valid URL.
- Test empty and invalid URL handling.
- Test download error handling.
- Test avatar replacement.
- Test directory deletion.
- Test non-actor post type protection.

obenland commented on includes/class-attachments.php:
```diff
@@ -933,4 +961,59 @@ private static function get_files_gallery_block( $files ) {
+	 *
+	 * @return string|false The local avatar URL on success, false on failure.
+	 */
+	public static function save_actor_avatar( $avatar_url, $actor_id ) {
```
```suggestion
	public static function save_actor_avatar( $actor_id, $avatar_url ) {
```
Let's swap the order to be consistent with other API functions that have the ID first.

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -677,10 +685,39 @@ public static function get_avatar_url( $id ) {
+		$avatar_url = $local_url ? $local_url : $remote_avatar_url;
 		\update_post_meta( $id, '_activitypub_avatar_url', \esc_url_raw( $avatar_url ) );
 
 		return $avatar_url;
```
If `cache_avatar()` was able to handle actor data arrays, this could be reduced to `return self::cache_avatar();`

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -677,10 +685,39 @@ public static function get_avatar_url( $id ) {
+	 * @param Actor $actor   The actor object.
+	 */
+	private static function cache_avatar( $post_id, $actor ) {
+		$remote_avatar_url = object_to_uri( $actor->get_icon() );
```
Let's add an `is_array()` check so `get_avatar_url()` can use it, too?

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -677,10 +685,39 @@ public static function get_avatar_url( $id ) {
+
+		// Store the local URL if caching succeeded, otherwise store the remote URL.
+		$avatar_url = $local_url ? $local_url : $remote_avatar_url;
+		\update_post_meta( $post_id, '_activitypub_avatar_url', \esc_url_raw( $avatar_url ) );
```
Let's return the stored avatar_url so `get_avatar_url()` can use it, too

obenland commented on includes/class-attachments.php:
```diff
@@ -27,18 +28,33 @@ class Attachments {
+	 *
+	 * @var int
+	 */
+	const MAX_AVATAR_DIMENSION = 100;
```
if we end up with dataviews for follow/following tables, we'd want this much bigger so avatars look good in Grid view. 512px? Not sure we need to worry about this at all

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -677,10 +685,39 @@ public static function get_avatar_url( $id ) {
+		$local_url = Attachments::save_actor_avatar( $remote_avatar_url, $post_id );
+
+		// Store the local URL if caching succeeded, otherwise store the remote URL.
+		$avatar_url = $local_url ? $local_url : $remote_avatar_url;
```
```suggestion
		$avatar_url = $local_url ?: $remote_avatar_url;
```

pfefferle committed:
Update includes/collection/class-remote-actors.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Simplify optimize_image and address PR feedback

- Return only file path from optimize_image instead of array with changed flag
- Let WordPress auto-detect mime type instead of manually setting it
- Add temp file extension handling for proper mime type detection
- Add get_unique_path helper to prevent file collisions during format conversion
- Move WP_Filesystem initialization to function start

pfefferle committed:
Address PR feedback for avatar caching

- Swap save_actor_avatar parameter order (actor_id first)
- Increase MAX_AVATAR_DIMENSION to 512px for better quality
- Add actor_id validation to prevent path traversal
- Add is_array check to cache_avatar for flexibility
- Return avatar_url from cache_avatar
- Use short ternary for avatar_url assignment
- Update tests to match new parameter order

obenland commented on includes/collection/class-remote-actors.php:
```diff
@@ -677,9 +685,39 @@ public static function get_avatar_url( $id ) {
+			$remote_avatar_url = object_to_uri( $actor['icon'] ?? ( $actor->icon ?? null ) );
+		} else {
+			$remote_avatar_url = object_to_uri( $actor->get_icon() );
+		}
```
```suggestion
		$data              = $actor instanceof Actor ? $actor->to_array() : (array) $actor;
		$remote_avatar_url = object_to_uri( $data['icon'] ?? null );
```

pfefferle committed:
Update includes/collection/class-remote-actors.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Add changelog entry

pfefferle committed:
Merge branch 'trunk' into add/avatar-caching

nickbohle commented:
Hi. Have you tested this with Automattic's Gravatar Enhanced? The plug-in caches not only Gravatars in comments, but also the Fediverse avatars.
